### PR TITLE
Require storage type and identifier

### PR DESCRIPTION
**Pull request by noseglid**
To avoid matching strings like `Something <= OtherThing` where the
`<` of the comparison operator was mistaken for a generic, require
there to be at least one storage type and at least one identifier.

Dollar are valid identifiers, added that in a few places it was
missing.

Also simplifies some regexes. E.g. `\w` matches `a-zA-Z0-9_` so
some redundant chars could be removed.

### CODE REVIEW

**Comment by noseglid** [comment]
Sample:
screen shot 2016-06-14 at 3 58 41 pm

File: grammars/java.cson [hunk 1]
```diff
@@ -1120,17 +1120,18 @@
           |
           ((\\w+\\.)*[A-Z]+\\w*) # e.g. `javax.ws.rs.Response`, or `String`
         )
-        ([\\w\\d<>\\[\\],?][\\w\\d_<>\\[\\],?\\s]*)?
+        ([\\w\\d<>\\[\\],][\\w\\d<>\\[\\],?\\s]*)?
```
Comment (MaximSokolov) [inline grammars/java.cson]: `\d` is also redundant as `\w` includes digits. Same for below

Comment (noseglid) [inline grammars/java.cson]: Right you are!

**Comment by 50Wliu** [comment]
@noseglid Can you add some regression tests for this and #58?  I noticed that for the previous two PRs none of the specs caught anything wrong.

**Comment by noseglid** [comment]
Sure. I've added some specs for most of the cases.

**Comment by 50Wliu** [comment]
Ran this through another quick eyeball using the same file that I've been seeing all the issues on and couldn't find anything wrong.  Here's hoping I didn't miss anything...

**Comment by noseglid** [comment]
Rather quick to deploy and quick to fix than slow to both :)
